pkg/image/controller/signature/signature_import_controller.go

```diff
@@ -0,0 +1,224 @@
+		s.queue.AddRateLimited(key)
+		return true
+	}
+
+	s.queue.Forget(key)
+	return true
+}
+
+func (s *SignatureImportController) enqueueImage(obj interface{}) {
+	_, ok := obj.(*imageapi.Image)
+	if !ok {
+		return
+	}
+	key, err := controller.KeyFunc(obj)
+	if err != nil {
+		glog.Errorf("Couldn't get key for object %+v: %v", obj, err)
+		return
+	}
+	s.queue.Add(key)
+}
+
+func (s *SignatureImportController) syncImageSignatures(key string) error {
+	glog.V(4).Infof("Initiating download of signatures for %s", key)
+	image, err := s.imageLister.Get(key)
+	if err != nil {
+		glog.V(4).Infof("Unable to get image %v: %v", key, err)
+		return err
+	}
+
+	currentSignatures, err := s.fetchImageSignaturesFn(image)
+	if err != nil {
+		glog.V(4).Infof("Failed to fetch image %s signatures: %v", image.Name, err)
+		return err
+	}
+
+	// Having no signatures means no-op (we don't remove stored signatures when
+	// the sig-store no longer have them).
+	if len(currentSignatures) == 0 {
+		glog.V(4).Infof("No signatures dowloaded for %s", image.Name)
+		return nil
+	}
+
+	t, err := kapi.Scheme.DeepCopy(image)
+	if err != nil {
+		return err
+	}
+	newImage := t.(*imageapi.Image)
+
+	shouldUpdate := false
+
+	// Only add new signatures, do not override existing stored signatures as that
+	// can void their verification status.
+	for _, c := range currentSignatures {
+		found := false
+		for _, s := range newImage.Signatures {
+			if s.Name == c.Name {
+				found = true
+				break
+			}
+		}
+		if !found {
+			newImage.Signatures = append(newImage.Signatures, c)
+			shouldUpdate = true
+		}
+	}
+
+	// Avoid unnecessary updates to images.
+	if !shouldUpdate {
+		return nil
+	}
+	glog.V(4).Infof("Image %s now has %d signatures", newImage.Name, len(newImage.Signatures))
+
+	if _, err := s.imageClient.Image().Images().Update(newImage); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (s *SignatureImportController) fetchImageSignatures(image *imageapi.Image) ([]imageapi.ImageSignature, error) {
+	reference, err := docker.ParseReference("//" + image.DockerImageReference)
+	if err != nil {
+		return nil, err
+	}
+	source, err := reference.NewImageSource(nil, nil)
+	if err != nil {
+		return nil, err
+	}
+	defer source.Close()
+
+	ctx, cancel := context.WithTimeout(s.ctx, s.fetchTimeout)
+	defer cancel()
+
+	signatures, err := source.GetSignatures(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	ret := []imageapi.ImageSignature{}
+	for _, blob := range signatures {
+		sig := imageapi.ImageSignature{Type: imageapi.ImageSignatureTypeAtomicImageV1}
+		// This will use the name of the image (sha256:xxxx) and the SHA256 of the
+		// signature itself as the signature name has to be unique for each
+		// signature.
+		sig.Name = image.Name + "@" + fmt.Sprintf("%x", sha256.Sum256(blob))
+		sig.Content = blob
+		sig.Annotations = map[string]string{
+			SignatureManagedAnnotation: "true",
+		}
+		sig.CreationTimestamp = metav1.Now()
```

Comment:
Doesn't the save method do this?  What about UIDs?

Reply:
@smarterclayton no, the timestamps are nil..